### DISCUSSION TITLE
Unable to extract `commonMain` resources in `commonTest` when running android unit tests in KMM app (missing JVM resource generation for `androidTest`?).

### DISCUSSION THREAD

**Discussion by darronschall** [icerockdev/moko-resources#393] [General]
Hello!

I'm using [moko-mvvm](https://github.com/icerockdev/moko-mvvm) in a KMM app to share common view model business logic between my Android and iOS app. I'm also using moko-resources.

I'm able to use my `commonMain/resources` in my Android and iOS UI code without issue, e.g.:

```swift
// Example resource usage in SwiftUI
Button(MR.strings().example_button_label_resource_string.desc().localized())
```

```kotlin
// Example resource usage in ComposeUI
Text(text = stringResource(id = MR.example_button_label_resource_string.resourceId))
```

The problem I'm having is that I would like to use a resource string in my viewModel. There is no helper for this out-of-the-box, so I created a platform wrapper, like this:

```kotlin
// in `commonMain`
interface Resources {
    fun string(stringResource: StringResource): String
    // TODO: More helpers here for substitutions, plurals, etc.
}
```

```kotlin
// in `iosMain`
class IosResources : Resources {
    override fun string(stringResource: StringResource): String {
        return stringResource.desc().localized()
    }
}
```

```kotlin
// in `androidMain`
class AndroidResources(
    private val context: Context
) : Resources {
    override fun string(stringResource: StringResource): String {
        return stringResource.desc().toString(context)
    }
}
```

I can then provide this wrapper to my view model. This lets me go through the `Resources` interface to extract a resource string that the viewModel can bind to. 

For example, given this resource (in `commonMain/resources/MR/base`):

```xml
<?xml version="1.0" encoding="UTF-8" ?>
<resources>
    <string name="example_error">Demo error message.</string>
</resources>
```

I can use my `Resources` interface to get the value at runtime in my view model:

```kotlin
// Supply `resources` via class constructor; e.g. `get()` from Koin after adding platform-specific `Resources`
// implementation to object graph.
class ExampleViewModel(private val resources: Resources) : ViewModel() {
    // Sample value for display in the view
    private val _errorMessage: CMutableStateFlow<String> = MutableStateFlow("").cMutableStateFlow()
    val errorMessage: CStateFlow<String> = _errorMessage.cStateFlow()  

    fun demoError() {
        // Use `resources.string` to extract the string value matching the resource key
        _errorMessage.value = resources.string(MR.strings.example_error)
    }
}
```

This works great. When running my Android and iOS app, I can see the "Demo error message" string in the view.

The problem I'm having is in my view model test declared in `commonTest`. I'm unable to extract the resources because `androidTest` runs on the JVM as unit tests, and _it doesn't look like moko-resources generates those resources_.

```kotlin
class ExampleViewModelTest {
    // This is not a great test, but trying to set up the minimal structure to demonstrate the problem
    @Test
    fun `error message is properly set on failure`() = runTest {
        val resources = createResources()
        val viewModel = ExampleViewModel(resources)
        
        viewModel.demoError()
        
        assertEquals("Demo error message.", viewModel.errorMessage.value)
    }
}
```

I made a `createResources` test helper, so that when running the tests I can use the correct implementation for my `Resources` wrapper. This type of test passes when running `iosSimulatorArm64Test` (the resources are available and work fine), but fails when running `testReleaseUnitTest` because the JVM resources cannot be found.

Here's what `createResources()` looks like:

```kotlin
// in `commonTest`
expect fun createResources(): Resources
```

```kotlin
// in `iosTest`
actual fun createResources(): Resources {
    return IosResources()
}
```

```kotlin
// in `androidTest
actual fun createResources(): Resources {
    // Can't do this, because no Android context (no instrumentation)
    // return AndroidResources(no_context_available_in_unit_tests)

    // So, we have to take another approach....
    // Since this runs via the JVM, try to load JVM resource instead:
    return object : Resources {
         override fun string(stringResource: StringResource): String {
             val resourcesClassLoader: ClassLoader = ClassLoader.getSystemClassLoader()
             // !!! This doesn't work; there are no resources available. Maybe I have the wrong bundle?
             val resourceBundle = resourcesClassLoader.getResourceBundle("resources", Locale.getDefault())
             return resourceBundle.getString(stringResource.resourceId.toString())
         }
     }
}

// Helper copied from https://github.com/icerockdev/moko-resources/blob/a4759a4439a492b6b8a6f7fa631be7cf0bdbe248/resources/src/jvmMain/kotlin/dev/icerock/moko/resources/ClassLoaderExt.kt#L10
 fun ClassLoader.getResourceBundle(bundleName: String, locale: Locale): ResourceBundle {
     return ResourceBundle.getBundle(
         bundleName,
         locale,
         this,
         // Otherwise, the default locale will be picked from system settings
         ResourceBundle.Control.getNoFallbackControl(ResourceBundle.Control.FORMAT_DEFAULT)
     )
 }
```

I tried to have my `androidTest` resource loader take the same kind of approach that moko-resources [`jvmMain`](https://github.com/icerockdev/moko-resources/tree/master/resources/src/jvmMain/kotlin/dev/icerock/moko/resources) does, but it's still not quite right.

I also tried using the `resources-test` library, but the [`createStringResourceMock()`](https://github.com/icerockdev/moko-resources/blob/master/resources-test/src/commonMain/kotlin/dev/icerock/moko/resources/test/MockResourcesFactory.kt#L12) does not help me pull the string value from the strings.xml file in my `commonTest` test.

In summary: 

 * How can I instruct moko-resources to generate JVM resources so that I can use them in `androidTest`? I don't want to add a JVM target to my project (it is iOS + Android only). This _seems_ like a bug that can be solved with a plugin update?
     * Or, is there a different approach to `androidTest` that I can take other than the ones I've already tried?

 * Related, would a future release of moko-resources be open to including the `Resources` interface and platform implementations, to make it easier to use resources in moko-mvvm view models like I demonstrated? I'm happy to submit a PR if others see value in it.
